bluesky/abc.py

```diff
@@ -0,0 +1,264 @@
+
+        They should contain the keys 'time', 'data', and 'timestamps'.
+        A 'uid' is added by the RunEngine.
+        """
+        ...
+
+    def describe_collect(self) -> Dict[str, Configuration]:
+        """This is like ``describe()`` on readable devices, but with an extra layer of nesting.
+
+        Since a flyer can potentially return more than one event stream, this is a dict
+        of stream names (strings) mapped to a ``describe()``-type output for each.
+        """
+        ...
+
+    # Remaining same as Readable
+    def read_configuration(self) -> Configuration:
+        """Same as for a Readable device."""
+        ...
+
+    def describe_configuration(self) -> Configuration:
+        """Same as for a Readable device."""
+        ...
+
+
+@runtime_checkable
+class Stageable(Protocol):
+    def stage(self) -> List[Any]:
+        """An optional hook for "setting up" the device for acquisition.
+
+        It should return a list of devices including itself and any other
+        devices that are staged as a result of staging this one.
+        (The ``parent`` attribute expresses this relationship: a device should
+        be staged/unstaged whenever its parent is staged/unstaged.)
+        """
+        ...
+
+    def unstage(self) -> List[Any]:
+        """A hook for "cleaning up" the device after acquisition.
+
+        It should return a list of devices including itself and any other
+        devices that are unstaged as a result of unstaging this one.
+        """
+        ...
+
+
+@runtime_checkable
+class Pausable(Protocol):
+    def pause(self) -> Union[NoReturn, None]:
+        """Perform device-specific work when the RunEngine pauses."""
+        ...
+
+    def resume(self) -> Union[NoReturn, None]:
+        """Perform device-specific work when the RunEngine resumes after a pause."""
+        ...
+
+
+@runtime_checkable
+class Stoppable(Protocol):
+    def stop(self, sucess=True) -> Union[NoReturn, None]:
+        """Safely stop a device that may or may not be in motion.
+
+        The argument ``success`` is a boolean.
+        When ``success`` is true, bluesky is stopping the device as planned
+        and the device should stop "normally".
+        When ``success`` is false, something has gone wrong and the device
+        may wish to take defensive action to make itself safe.
+        """
+        ...
+
+
+@runtime_checkable
+class Subscribable(Protocol):
+    def subscribe(self, function: Callable) -> Union[NoReturn, None]:
+        """Subscribe to updates in value of a device.
+
+        When the device has a new value ready, it should call ``function``
+        asynchronously in a separate thread.
+
+        Needed for :doc:`monitored <async>`.
+        """
+        ...
+
+    def clear_sub(self, function: Callable) -> Union[NoReturn, None]:
+        """Remove a subscription."""
+        ...
+
+
+@runtime_checkable
+class Checkable(Protocol):
+    def check_value(self, value: Any) -> Union[NoReturn, None]:
```

Comment:
@thomascobb @tacaswell Do you recall what we said we might do about this? Would this be superseded by `validate`?